### PR TITLE
fix!: fix regression caused by unbalanced partitions and splitting ranges

### PR DESCRIPTION
**Pull request by evenyag**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## Refer to a related PR or issue link (optional)
- https://github.com/GreptimeTeam/greptimedb/issues/5052
- https://github.com/GreptimeTeam/greptimedb/issues/4828

## What's changed and what's your intention?

Fix several issues in SeqScan related to #5052
- Avoids pushing row group indices before split parts. It may cause the merge reader to merge all row groups of the same file, which is unexpected
    - **Before this PR, compaction will load all row groups into memory.** This is a regression introduced in https://github.com/GreptimeTeam/greptimedb/pull/4809
- Adds the number of row groups to the `SourceIndex` so we can know how to split the source.
- Makes the number of rows in partitions close to each other if possible and doesn't always create `target_partitions` partitions so the query engine can repartition inputs.

## Checklist

- [ ] I have written the necessary rustdoc comments.
- [ ] I have added the necessary unit tests and integration tests.
- [x] This PR requires documentation updates.

### CODE REVIEW

File: config/datanode.example.toml [hunk 1]
```diff
@@ -492,12 +492,6 @@ experimental_write_cache_ttl = "8h"
 ## Buffer size for SST writing.
 sst_write_buffer_size = "8MB"
 
-## Parallelism to scan a region (default: 1/4 of cpu cores).
```
Comment (killme2008) [inline config/datanode.example.toml:495]: So it's another breaking change.

File: src/query/src/optimizer/parallelize_scan.rs [hunk 2]
```diff
@@ -109,21 +110,67 @@ impl ParallelizeScan {
...
+        mut ranges: Vec<PartitionRange>,
         expected_partition_num: usize,
     ) -> Vec<Vec<PartitionRange>> {
-        let actual_partition_num = expected_partition_num.min(ranges.len()).max(1);
+        if ranges.is_empty() {
+            // Returns a single partition with no range.
+            return vec![vec![]];
+        }
+
+        // Sort ranges by number of rows in descending order.
```
Comment (killme2008) [inline src/query/src/optimizer/parallelize_scan.rs:127]: Consider implementing a fast path for `ranges.len() == 1` to skip the costly calculation below.

File: src/mito2/src/read/seq_scan.rs [hunk 3]
```diff
@@ -112,23 +105,16 @@ impl SeqScan {
...
-        )
-        .await?;
+        let reader =
+            Self::build_all_merge_reader(&self.stream_ctx, partition_ranges, &part_metrics).await?;
         Ok(Box::new(reader))
     }
 
     /// Builds a merge reader that reads all data.
+    /// Callers MUST not split ranges before calling this method.
     async fn build_all_merge_reader(
```
Comment (killme2008) [inline src/mito2/src/read/seq_scan.rs:115]: Is it only used for compaction reader? If so , i think it must be renamed too.

File: src/mito2/src/read/scan_region.rs [hunk 4]
```diff
@@ -104,6 +96,17 @@ impl Scanner {
             Scanner::Unordered(unordered_scan) => unordered_scan.input().file_ids(),
         }
     }
+
+    /// Sets the target partitions for the scanner. It can controls the parallelism of the scanner.
+    pub(crate) fn set_target_partitions(&mut self, target_partitions: usize) {
+        use store_api::region_engine::{PrepareRequest, RegionScanner};
```
Comment (killme2008) [inline src/mito2/src/read/scan_region.rs:102]: Move the `use` out of the function.

File: src/query/src/optimizer/parallelize_scan.rs (hunk 2 above)
Comment (fengjiachun) [reply src/query/src/optimizer/parallelize_scan.rs:127]: The `sort_by` function also takes this internally (fast path)

File: src/mito2/src/read/scan_region.rs (hunk 4 above)
Comment (evenyag) [reply src/mito2/src/read/scan_region.rs:102]: This method is under `#[cfg(test)]` so moving the `use` to outside will cause `unused` warning.

File: config/datanode.example.toml (hunk 1 above)
Comment (evenyag) [reply config/datanode.example.toml:495]: Yes. It's still compatible with the old toml file.

**Review by killme2008 — APPROVED** [review]
LGTM

**Pull request merged**